### PR TITLE
Added text stroking

### LINKED ISSUE #2209: ImageFont/ImageDraw should support text stroking
Hey,

Something that I find myself doing quite often when drawing text to an image is to stroke the text with a high-contrast outline.
I find this helps improve readability over cluttered backgrounds.

Neither pillow.ImageDraw nor pillow.ImageFont currently includes a function while allows ImageDraw to draw stroked text with a specific color outline.

It'd be pretty cool if ImageDraw.text() would include two additional method parameters.

- One to indicate the size of the outline in pixels
- One to indicate the colour of the outline (similar to the current fill parameter)

For example:

``draw.text((10, 10), "hello", font=font, outline=BLACK, outline_size=2)``

### PR DESCRIPTION
**Pull request by radarhere**
Resolves #2209. Replacement for #2224

Adds `strokeWidth` and `strokeFill` arguments, such that -

```python
from PIL import Image, ImageDraw, ImageFont

font = ImageFont.truetype("Tests/fonts/FreeMono.ttf", 40)
font.getsize_multiline("A", strokeWidth=2)
font.getsize("ABC\nAaaa", strokeWidth=2)

im = Image.new("RGB", (100, 100))
draw = ImageDraw.Draw(im)
draw.textsize("A", font, strokeWidth=2)
draw.multiline_textsize("ABC\nAaaa", font, strokeWidth=2)
draw.text((10, 10), "A", "#f00", font, strokeWidth=2, strokeFill="#0f0")
draw.multiline_text((10, 10), "A\nB", "#f00", font, strokeWidth=2, strokeFill="#0f0")
```

Uses FreeType's stroking. See https://www.freetype.org/freetype2/docs/reference/ft2-glyph_stroker.html for more information, and part of https://www.freetype.org/freetype2/docs/tutorial/example2.cpp for example code.

Also renames a FT_GlyphSlot variable from 'glyph' to 'glyphSlot'.

### CODE REVIEW

File: src/PIL/ImageDraw.py [hunk 1]
```diff
@@ -261,24 +261,63 @@ def _multiline_split(self, text):
...
+        fill=None,
+        font=None,
+        anchor=None,
+        strokeWidth=0,
+        strokeFill=None,
+        *args,
+        **kwargs
+    ):
         if self._multiline_check(text):
             return self.multiline_text(xy, text, fill, font, anchor, *args, **kwargs)
```
Comment (nulano) [inline src/PIL/ImageDraw.py]: This should pass the `strokeWidth` and `strokeFill` parameters, no?

Comment (radarhere) [reply src/PIL/ImageDraw.py]: Good point, thanks. I've updated the commit.

File: src/PIL/ImageDraw.py [hunk 2]
```diff
@@ -261,24 +261,95 @@ def _multiline_split(self, text):
...
+        fill=None,
+        font=None,
+        anchor=None,
+        spacing=4,
+        align="left",
+        direction=None,
+        features=None,
+        language=None,
+        strokeWidth=0,
+        strokeFill=None,
```
Comment (hugovk) [inline src/PIL/ImageDraw.py]: Any particular reason for stroke width and fill params to be in this order (and in `multiline_text`)?

If not, how about alphabetical order?

File: src/PIL/ImageDraw.py [hunk 3]
```diff
@@ -261,24 +261,95 @@ def _multiline_split(self, text):
...
                 except TypeError:
                     mask = font.getmask(text)
-            self.draw.draw_bitmap(xy, mask, ink)
+            if strokeOffset:
+                coord = coord[0] + strokeOffset[0], coord[1] + strokeOffset[1]
+            self.draw.draw_bitmap(coord, mask, ink)
+
+        ink = getink(fill)
+        if ink is not None:
+            strokeInk = None
```
Comment (hugovk) [inline src/PIL/ImageDraw.py]: `snake_case`?
```suggestion
            stroke_ink = None
```

File: src/_imagingft.c [hunk 4]
```diff
@@ -790,7 +791,13 @@ font_render(FontObject* self, PyObject* args)
     int index, error, ascender, horizontal_dir;
     int load_flags;
     unsigned char *source;
-    FT_GlyphSlot glyph;
+    FT_Glyph glyph;
+    FT_GlyphSlot glyphSlot;
+    FT_Bitmap bitmap;
+    FT_BitmapGlyph bitmapGlyph;
+    int strokeWidth = 0;
```
Comment (hugovk) [inline src/_imagingft.c]:
```suggestion
    int stroke_width = 0;
```

File: src/_imagingft.c [hunk 5]
```diff
@@ -790,7 +791,13 @@ font_render(FontObject* self, PyObject* args)
     int index, error, ascender, horizontal_dir;
     int load_flags;
     unsigned char *source;
-    FT_GlyphSlot glyph;
+    FT_Glyph glyph;
+    FT_GlyphSlot glyphSlot;
+    FT_Bitmap bitmap;
+    FT_BitmapGlyph bitmapGlyph;
```
Comment (hugovk) [inline src/_imagingft.c]:
```suggestion
    FT_BitmapGlyph bitmap_glyph;
```

File: src/_imagingft.c [hunk 6]
```diff
@@ -790,7 +791,13 @@ font_render(FontObject* self, PyObject* args)
     int index, error, ascender, horizontal_dir;
     int load_flags;
     unsigned char *source;
-    FT_GlyphSlot glyph;
+    FT_Glyph glyph;
+    FT_GlyphSlot glyphSlot;
```
Comment (hugovk) [inline src/_imagingft.c]: Be consistent with the other variables, use `snake_case`:
```suggestion
    FT_GlyphSlot glyph_slot;
```

File: src/_imagingft.c [hunk 7]
```diff
@@ -819,9 +827,19 @@ font_render(FontObject* self, PyObject* args)
         Py_RETURN_NONE;
     }
 
+    if (strokeWidth) {
+        error = FT_Stroker_New(library, &stroker);
+        if (error)
```
Comment (hugovk) [inline src/_imagingft.c]: Let's defensively use parentheses, even for single-line `if`s

File: src/_imagingft.c [hunk 8]
```diff
@@ -819,9 +827,19 @@ font_render(FontObject* self, PyObject* args)
...
+            return geterror(error);
+
+        FT_Stroker_Set(stroker, (FT_Fixed)strokeWidth*64, FT_STROKER_LINECAP_ROUND, FT_STROKER_LINEJOIN_ROUND, 0);
+    }
+
     im = (Imaging) id;
     /* Note: bitmap fonts within ttf fonts do not work, see #891/pr#960 */
-    load_flags = FT_LOAD_RENDER|FT_LOAD_NO_BITMAP;
+    load_flags = FT_LOAD_NO_BITMAP;
+    if (stroker == NULL)
```
Comment (hugovk) [inline src/_imagingft.c]: And here

File: src/_imagingft.c [hunk 9]
```diff
@@ -847,34 +867,57 @@ font_render(FontObject* self, PyObject* args)
...
+        if (stroker != NULL) {
+            error = FT_Get_Glyph(glyphSlot, &glyph);
+            if (!error) {
+                error = FT_Glyph_Stroke(&glyph, stroker, 1);
+            }
+            if (!error) {
+                FT_Vector origin = {0, 0};
+                error = FT_Glyph_To_Bitmap(&glyph, FT_RENDER_MODE_NORMAL, &origin, 1);
+            }
+            if (error)
```
Comment (hugovk) [inline src/_imagingft.c]: And here

File: docs/reference/ImageDraw.rst [hunk 10]
```diff
@@ -297,6 +297,15 @@ Methods
 
                      .. versionadded:: 6.0.0
 
+    : param strokeWidth: The width of the text stroke.
```
[the remark this answers is not in the document]
Comment (radarhere) [reply docs/reference/ImageDraw.rst]: Okay, I've updated the commit, and all snake case situations should now be addressed. Let me know if not.

File: src/_imagingft.c (hunk 7 above)
Comment (radarhere) [reply src/_imagingft.c]: All single-line `if`s should now be using parentheses.

File: src/PIL/ImageDraw.py (hunk 2 above)
Comment (radarhere) [reply src/PIL/ImageDraw.py]: The only reason would be that stroke fill has no meaning without stroke width, making stroke fill optional compared to stroke width - in the situation where a user wanted to stroke with the same fill as the text, they would specify width, but leave fill as the default.

If you think it's better to easily remember the arguments as being in alphabetical order though, then I can make that change.

File: docs/reference/ImageDraw.rst [hunk 11]
```diff
@@ -297,6 +297,15 @@ Methods
...
 
+    :param stroke_width: The width of the text stroke.
+
+                     .. versionadded:: 6.2.0
+
+    :param stroke_fill: Color to use for the text stroke. If not given, will default to
+                        the ``fill`` parameter.
+
+                     .. versionadded:: 6.2.0
+
```
Comment (nulano) [inline docs/reference/ImageDraw.rst]: These should probably be added to the method signature above.

File: docs/reference/ImageDraw.rst [hunk 12]
```diff
@@ -372,7 +381,11 @@ Methods
 
                      .. versionadded:: 6.0.0
 
-.. py:method:: PIL.ImageDraw.ImageDraw.multiline_textsize(text, font=None, spacing=4, direction=None, features=None, language=None)
+    :param stroke_width: The width of the text stroke.
+
+                     .. versionadded:: 6.2.0
```
Comment (nulano) [inline docs/reference/ImageDraw.rst]: Same here.

File: docs/reference/ImageDraw.rst (hunk 11 above)
Comment (radarhere) [reply docs/reference/ImageDraw.rst]: Thanks, done

File: docs/reference/ImageDraw.rst (hunk 12 above)
Comment (radarhere) [reply docs/reference/ImageDraw.rst]: Done

File: src/PIL/ImageDraw.py [hunk 13]
```diff
@@ -261,24 +261,95 @@ def _multiline_split(self, text):
...
-            ink = fill
-        if ink is not None:
+
+        def getink(fill):
+            ink, fill = self._getink(fill)
+            if ink is None:
+                return fill
+            return ink
+
+        def drawText(ink, stroke_width=0, stroke_offset=None):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py]:
```suggestion
        def draw_text(ink, stroke_width=0, stroke_offset=None):
```

File: src/PIL/ImageDraw.py [hunk 14]
```diff
@@ -261,24 +261,95 @@ def _multiline_split(self, text):
...
+
+        ink = getink(fill)
+        if ink is not None:
+            stroke_ink = None
+            if stroke_width:
+                stroke_ink = getink(stroke_fill) if stroke_fill is not None else ink
+
+            if stroke_ink is not None:
+                # Draw stroked text
+                drawText(stroke_ink, stroke_width)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py]:
```suggestion
                draw_text(stroke_ink, stroke_width)
```

File: src/PIL/ImageDraw.py [hunk 15]
```diff
@@ -261,24 +261,95 @@ def _multiline_split(self, text):
...
+
+                # Draw normal text
+                drawText(ink, 0, (stroke_width, stroke_width))
```
Comment (hugovk) [inline src/PIL/ImageDraw.py]:
```suggestion
                draw_text(ink, 0, (stroke_width, stroke_width))
```

File: src/PIL/ImageDraw.py [hunk 16]
```diff
@@ -261,24 +261,95 @@ def _multiline_split(self, text):
...
+            else:
+                # Only draw normal text
+                drawText(ink)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py]:
```suggestion
                draw_text(ink)
```

**Pull request merged**